libraries/rodeos/include/b1/rodeos/callbacks/compiler_builtins.hpp

```diff
@@ -0,0 +1,397 @@
+#pragma once
+
+#include <b1/rodeos/callbacks/vm_types.hpp>
+#include <compiler_builtins.hpp>
+#include <softfloat.hpp>
+
+namespace b1::rodeos
+{
+   template <typename Derived>
+   struct compiler_builtins_callbacks
+   {
+      Derived& derived() { return static_cast<Derived&>(*this); }
+
+      void __ashlti3(legacy_ptr<__int128> ret, uint64_t low, uint64_t high, uint32_t shift)
+      {
+         if (shift >= 128)
+         {
+            *ret = 0;
+         }
+         else
+         {
+            unsigned __int128 i = high;
+            i <<= 64;
+            i |= low;
+            i <<= shift;
+            *ret = (__int128)i;
+         }
+      }
+
+      void __ashrti3(legacy_ptr<__int128> ret, uint64_t low, uint64_t high, uint32_t shift)
+      {
+         // retain the signedness
+         *ret = high;
+         *ret <<= 64;
+         *ret |= low;
+         *ret >>= shift;
```

Comment:
Oh boy. eosio 2.0 has that identical code. I'll switch the implementations of all these functions to be identical to 2.0 and add UB TODOs with a note about 2.0 also needing analysis and fixes to prevent consensus breaks.